server/src/test/java/org/elasticsearch/cluster/coordination/JoinTaskExecutorTests.java

```diff
@@ -488,6 +499,92 @@ public void testIgnoresOlderTerms() throws Exception {
         );
     }
 
+    public void testDesiredNodesMembershipIsUpgradedWhenNewNodesJoin() throws Exception {
+        final var allocationService = createAllocationService();
+        final RerouteService rerouteService = (reason, priority, listener) -> listener.onResponse(null);
+
+        final var joinTaskExecutor = new JoinTaskExecutor(allocationService, rerouteService);
+
+        final var actualizedDesiredNodes = randomList(0, 5, this::createActualizedDesiredNode);
+        final var pendingDesiredNodes = randomList(0, 5, this::createPendingDesiredNode);
+        final var joiningDesiredNodes = randomList(1, 5, this::createPendingDesiredNode);
+
+        final var masterNode = newDiscoveryNode("master");
+        final var discoveryNodes = DiscoveryNodes.builder()
+            .add(masterNode)
+            .localNodeId(masterNode.getId())
+            .masterNodeId(masterNode.getId());
+
+        for (DesiredNodeWithStatus actualizedDesiredNode : actualizedDesiredNodes) {
+            discoveryNodes.add(newDiscoveryNode(actualizedDesiredNode.externalId()));
+        }
+
+        final List<DiscoveryNode> joiningNodes = joiningDesiredNodes.stream()
+            .map(desiredNode -> newDiscoveryNode(desiredNode.externalId()))
+            .toList();
+
+        // Add some nodes in the cluster that are not part of the desired nodes
+        for (int i = 0; i < randomInt(5); i++) {
```

Comment:
This gives you a higher probability for lower numbers, since the `randomInt` is reevaluated for every iteration, you probably meant to take the number into a variable first? Or use `IntStream.range` instead.